SI-6541 valid wildcard existentials for case-module-unapply

lrytz opened the pull request:
Instead of letting the compiler infer the return type of case module
unapply methods, provide them explicitly.

lrytz committed:
SI-6541 valid wildcard existentials for case-module-unapply

Instead of letting the compiler infer the return type of case module
unapply methods, provide them explicitly.

This is enabled only under -Xsource:2.12, because the change is not
source compatible.

lrytz commented:
Review by @retronym

retronym commented on src/compiler/scala/tools/nsc/typechecker/Unapplies.scala:
```diff
@@ -142,17 +142,30 @@ trait Unapplies extends ast.TreeDSL {
+          gen.rootScalaDot(tpnme.Boolean)
+        case params :: _ =>
+          val constrParamTypes = params.map(param => repeatedToSeq(param.tpt))
+          AppliedTypeTree(gen.rootScalaDot(tpnme.Option), List(treeBuilder.makeTupleType(constrParamTypes)))
```
This ain't hygenic.

``` scala
case class C2(b: Tuple1[_ <: String])
object C2 {
  class String
}
```

``` noformat
test/files/run/t6541.scala:28: error: type mismatch;
 found   : (_$1(in method apply),) where type _$1(in method apply) <: C2.String
 required: Tuple1[_ <: String]
case class C2(b: Tuple1[_ <: String])
           ^
```

retronym replied:
[Last time I tried to fix this](https://github.com/retronym/scala/compare/ticket/6541), I ended up with a custom type completer for `unapply` to add enough laziness to the mix so we could peek at the `info` of the case class to derive the type of the `unapply` method.

lrytz replied:
You are right about hygiene. But note that the error message talks about `apply` (not `unapply`), so this failure is not introduced by my change. Arguably, the failure is according to the spec: http://www.scala-lang.org/files/archive/spec/2.11/05-classes-and-objects.html#case-classes.

We should do the same for the parameter types of apply and the return type of unapply: both hygienic, or both not.

We should also include `copy` in the discussion – it's not systematic that it works differently. The problem with copy is that we have a stronger argument for making it hygienic: otherwise we can get a failure even without an explicit shadowing type declaration:

``` scala
class A {
  class C
  case class B(x: C) extends A { def copy(x: C = this.x) = ... }
}
```

The class parameter type is `A.this.C`. If copy was non-hygienic, the parameter type of `copy` would be `B.this.C` (since B extends A), so it would not compile.

retronym replied:
Right you are. In fact, the type parameters of `copy` are also prone to shadowing.

``` scala
case class C2[T <: String]() {
  class String
}
```

```
% scalac-hash v2.11.2 test/files/run/t6541.scala
test/files/run/t6541.scala:28: error: type arguments [T] do not conform to class C2's type parameter bounds [T <: String]
case class C2[T <: String]() {
           ^
one error found
```

In this case, we can defer that mess to a new ticket and proceed with this one. Perhaps though it needs to be under `-Xsource:2.12` as it changes signatures (albeit in a binary compatible way).

lrytz replied:
I logged https://issues.scala-lang.org/browse/SI-8884

Good point about compatibility, I only thought about binary.

I'm not sure if `-Xsource:2.12` is the right solution here, maybe better submit the PR against 2.12? Because the generated pickle is different, so in some sense it's binary incompatible. If you turn it on to compile a library, then a client might fail to compile against the new binary.

retronym replied:
I can't see a way that enabling this under `-Xsource:2.12` would lead to breakage under separate compilation with different flags. The type of the `unapply` is computed once and stored in the pickle.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Unapplies.scala:
```diff
@@ -142,17 +142,30 @@ trait Unapplies extends ast.TreeDSL {
+        case AppliedTypeTree(Select(_, tpnme.REPEATED_PARAM_CLASS_NAME), tps) => AppliedTypeTree(gen.rootScalaDot(tpnme.Seq), tps)
+        case _                                                                => tp
+      }
+      constrParamss(cdef) match {
```
- [x] `constrParamss` returns duplicated trees

retronym replied:
But needn't call it twice (once to determine the method name, and again here).

lrytz replied:
yep, thanks!

retronym commented:
Not sure what style we should settle on, but we might consider referring to `Option` and friends symfully rather that with the root qualified path.

e.g. https://github.com/retronym/scala/compare/review/4017/symful?expand=1

We don't need to change this PR, but let's give it some thought and seek some consistency in how we refer to things from the parser. /cc @densh

densh commented:
As far as I know the reason we have both root qualified and symbol qualified identifiers emitted by the parser due to the fact that symbol-only approach used to cause cyclic references while bootstrapping standard library. Maybe this is not an issue any more but it needs some investigation.

lrytz committed:
Fix default value for ScalaVersionSetting

The default value was NoScalaVersion before, because tryToSet (where
the default was supposed to be set) is not called at all if the option
is not specified.

The initial value of Xmigration is set to NoScalaVersion (which it was
before, the AnyScalaVersion argument was ignored). AnyScalaVersion
would be wrong, it would give a warning in `Map(1 -> "eis").values`
if the option was not specified. See tests.

lrytz commented:
The change is now under `-Xsource:2.12`, re-review by @retronym

retronym commented:
Great! One last thing: can you please add a `pos` test _without_ the `-Xsource:2.12` flag to back up the fact that: `val o: Option[Class[T]] forSome { type T } = C.unapply(C(classOf[String]))` used to work.

lrytz commented:
@retronym turned it into a repl test

retronym commented:
LGTM
